### PR TITLE
[primTorch] Make sure refs are exported via `__all__`

### PR DESCRIPTION
**Pull request by nkaretnikov**

### CODE REVIEW

**Comment by nkaretnikov** [comment]
noticed some names are not exported. it's easy to forget, so added a test as well. happy to change if there's a better way to test for this

**Comment by Lezcano** [comment]
Isn't there already a test for this in `test/allowlist_for_publicAPI.json`? Wouldn't it be better to use that test to run on these files?

**Comment by nkaretnikov** [comment]
@Lezcano 

> Isn't there already a test for this in test/allowlist_for_publicAPI.json? Wouldn't it be better to use that test to run on these files?

i think it's better to have a separate test for refs for now. the one you mention is used for public apis, so no modules with `_`, like `_refs`, should use that one

**Comment by nkaretnikov** [comment]
@Lezcano had to move from decomps, otherwise was getting duplicate def'n errors.

note: this also removes promotion wrappers, which were in decomps (but not in refs).

what promotion behavior do we want here?

- `vector_norm`, which is used by `norm` (and was used by decomp) has no promotion wrapper defined, so default?
- `logsumexp` calls `log` which has `INT_TO_FLOAT`, so we probably want it in `logsumexp` as well?

thoughts?

File: torch/testing/_internal/opinfo/refs.py [hunk 1]
```diff
@@ -113,6 +113,23 @@ def __init__(
...
+            # make sure name is exported via __all__
+            import torch
+
+            lst = name.split(".")
+            mod_name = lst[:-1]
+            func_name = lst[-1]
+
+            mod = torch
+            for i in range(len(mod_name)):
+                mod = getattr(mod, mod_name[i])
```
Comment (Lezcano) [inline torch/testing/_internal/opinfo/refs.py:126]: This is a hack. You want to do
```
mod = __import__(mod_name)
```
and remove the `import torch` above.

File: torch/testing/_internal/opinfo/refs.py [hunk 2]
```diff
@@ -113,6 +113,23 @@ def __init__(
...
+
+            if hasattr(mod, "__all__"):
```
Comment (Lezcano) [inline torch/testing/_internal/opinfo/refs.py:128]: What's a case of a module that doesn't have the `__all__` attribute and why?

File: torch/_decomp/decompositions.py [hunk 3]
```diff
@@ -1575,21 +1561,6 @@ def log_sigmoid_forward(self: Tensor) -> Tuple[Tensor, Tensor]:
...
-def norm(
-    self: Tensor,
-    p: Optional[float] = None,
-    dim: List[int] = None,
-    keepdim: bool = False,
-    dtype: Optional[torch.dtype] = None,
-):
-    if p is None:
-        p = 2.0
-    return torch.linalg.vector_norm(self, p, dim, keepdim, dtype=dtype)
```
[the remark this answers is not in the document]
Comment (nkaretnikov) [reply torch/_decomp/decompositions.py:1590]: decomp tests broke locally when i exported refs with explicit decompositions via __all__

File: torch/testing/_internal/opinfo/refs.py (hunk 1 above)
Comment (nkaretnikov) [reply torch/testing/_internal/opinfo/refs.py:126]: i had that initially (also tried with importlib). these don't work for

```
    PythonRefInfo(
        "ops.nvprims.var_mean",
```

which is why i have the above.

alternatively, we could skip this module instead and use the proper import mechanism.

File: torch/testing/_internal/opinfo/refs.py (hunk 2 above)
Comment (nkaretnikov) [reply torch/testing/_internal/opinfo/refs.py:128]: ops.nvprims.var_mean, see above

File: torch/testing/_internal/opinfo/refs.py (hunk 1 above)
Comment (Lezcano) [reply torch/testing/_internal/opinfo/refs.py:126]: @IvanYashchuk do you happen to know what's happening here?

File: torch/_decomp/decompositions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/_decomp/decompositions.py:1590]: That's an example where ghstack would have helped. You could have stacked a PR on top of this one with this fix.

File: torch/testing/_internal/opinfo/refs.py (hunk 1 above)
Comment (IvanYashchuk) [reply torch/testing/_internal/opinfo/refs.py:126]: `torch.ops.nvprims.var_mean` (and all other `torch.ops` modules) is not created until `torch._prims` is imported.

Comment (Lezcano) [reply torch/testing/_internal/opinfo/refs.py:126]: Also, it is not clear to me that this is the right place for this check. Why not making it a test in `test_ops`?

Comment (nkaretnikov) [reply torch/testing/_internal/opinfo/refs.py]: i've just added a comment.
- importing `_prims` didn't work
- `__import__` doesn't work correctly with `__all__`: for a `torch` submodule, it just returns `torch` as the module
- `importlib.import_module` works, but not with `nvprims`, as explained above

**Pull request closed without merging**